### ISSUE TITLE
proc-to-interface conversion clones String based objects

### ISSUE THREAD

**Issue by klobuczek** [jruby/jruby#5603]
### Environment

* jruby 9.2.5.0 (2.5.0) 2018-12-06 6d5a228 Java HotSpot(TM) 64-Bit Server VM 25.191-b12 on 1.8.0_191-b12 +jit [darwin-x86_64]
* Operating system and platform (e.g. `uname -a`)
Darwin berlin 18.2.0 Darwin Kernel Version 18.2.0: Mon Nov 12 20:24:46 PST 2018; root:xnu-4903.231.4~2/RELEASE_X86_64 x86_64

### Expected Behavior

Proc-to-interface conversion should preserve the object returned by the ruby block

### Actual Behavior

If the ruby block returns a String or any subclass of String the interface proxy returns a copy of the underlying object as String loosing the actual type. The behavior is correct for any other type.

Example:
```
public interface Interface<T>
{
    T execute();
}

public class TypeDemo {
    public static <T> T demo(Interface<T> i)
    {
        return i.execute();
    }
}
```
Then the following ruby script
```
class Abc; end
class AbcSubclass < Abc; end
class StringSubclas < String; end
class StringSubSubclas < StringSubclas; end
class HashSubclass < Hash; end
class ArraySubclass < Array; end

[1, "abc", Abc.new, AbcSubclass.new, StringSubclas.new, StringSubSubclas.new, HashSubclass.new, ArraySubclass.new, Class.new, [1, 2, 3], ["abc"], {a: 1}].each do |before|
  after = Java::TypeDemo.demo { before }
  puts "before.class: #{before.class}, after.class: #{after.class}, same objects: #{before.object_id == after.object_id}"
end
```
produces the following output:
```
before.class: Integer, after.class: Integer, same objects: true
before.class: String, after.class: String, same objects: false
before.class: Abc, after.class: Abc, same objects: true
before.class: AbcSubclass, after.class: AbcSubclass, same objects: true
before.class: StringSubclas, after.class: String, same objects: false
before.class: StringSubSubclas, after.class: String, same objects: false
before.class: HashSubclass, after.class: HashSubclass, same objects: true
before.class: ArraySubclass, after.class: ArraySubclass, same objects: true
before.class: Class, after.class: Class, same objects: true
before.class: Array, after.class: Array, same objects: true
before.class: Array, after.class: Array, same objects: true
before.class: Hash, after.class: Hash, same objects: true
```

**Comment by klobuczek** [comment]
Verified. Issue persists with:
```
jruby 9.2.6.0 (2.5.3) 2019-02-11 15ba00b Java HotSpot(TM) 64-Bit Server VM 25.191-b12 on 1.8.0_191-b12 +jit [darwin-x86_64]```

**Comment by lkalwa** [comment]
I recently encountered the same issue. 
I can try to find a fix for that, @headius any hint where I should start digging?

**Comment by kares** [comment]
believe this is "normal" for bare String -> since JRuby always converts Ruby -> Java by default.
here the target return type is `java.lang.Object` thus it does `RubyString#toJava(Object.class)`
however the sub-string path is weird and also that Integers seemingly no longer convert back-and-forth.
will take a look.

**Comment by kares** [comment]
yeah, its definitely that - and Integers also get a round-trip going `toJava` and than back to Ruby land.
one can confirm that by using a "bignum" Integer e.g. `100**20` - you won't get the same object back.
with "fixnum" `Integer` it seemingly seems to be the case but (except for the fixnum cache) their not the same Java objects ... its the way object_id gets calculated that they appear to be the same.

there isn't any fix for a patch release as it would like break existing scripts using Java Integration. 
we could change JI to NOT convert by default in 9.3 ... or at least have an option to turn it off. 
you could avoid the (default) conversion by using a JRuby specific return type `IRubyObject`

**Comment by klobuczek** [comment]
@kares found another type that is being converted to Java:
```before.class: Time, after.class: Java::JavaUtil::Date, same objects: false```
BTW the workaround by using `IRubyObject` does not work for me as I am implementing a gem for both jruby and MRI so the APIs must be identical.

**Comment by headius** [comment]
@klobuczek Going through Java integration there's currently no way to prevent the coercion to a Java type and back to Ruby; on the Java side, if you're calling an API that wants a string, we have to pass it a string.

We have considered adding more knobs and features to Java integration to allow you to control that coercion directly, but there has never been any consensus on what that ought to look like.

An alternative way to do what you want is to implement a small part of your library as a JRuby extension (in Java or another JVM language) and use that to interact with the rest of the JVM world. That will allow you to pass the IRubyObject directly through to the JRuby ext, and from there you can do what you wish.

You say that using IRubyObject did not work for you...do you have an example of what you tried?

**Comment by klobuczek** [comment]
@headius I never fully understood how to use IRubyObject for this problem, but I found another workaround by overriding the affected methods and wrapping the result of the block into a struct and unwrapping it when receiving back from Java. That way my objects are not converted to Java. Is there a nicer way by using IRubyObject without the need to create a struct?

```
        # work around jruby issue https://github.com/jruby/jruby/issues/5603
        Struct.new('Wrapper', :object)

        def write_transaction
          super { |tx| Struct::Wrapper.new(yield(tx)) }.object
        end

        # end work around
```

**Comment by headius** [comment]
Unfortunately I don't think there's a workaround for the low-level types like String, Integer, and so on that have equivalents in Java. We always look at the natural type of those values and convert them to Java equivalents, even when passed for an "Object" target parameter.

@kares I wonder...perhaps we can try softening that so that when you ask for "Object" we don't coerce to the equivalent Java type? It would break some code, to be sure, but most of the time if you're calling an API that wants a String it asks for a String. We'd still convert for that case, or for numeric cases calling parameter types of Number or Integer.

**Comment by headius** [comment]
Ok so the aggressive approach of removing auto-coercion for `Object` (and `Comparable` in this case) does not work. The main problem is that Java collections all get erased to `Object` so suddenly we start mixing `RubyString` into collections that clearly are intended for `String`.

But taking a step back...using IRubyObject *does* appear to work. RubyString does not autocoerce if the target type is IRubyObject; that falls back to RubyBasicObject, which would see that IRubyObject is a supertype of all Ruby objects and return them as-is. For example:

java:
```java
import org.jruby.runtime.builtin.IRubyObject;

public class Blah {
  public interface Foo { IRubyObject foo(); }
  public static IRubyObject blah(Foo foo) { return foo.foo(); }
}
```
ruby:
```ruby
class MyFoo
  include Java::Blah::Foo
  def foo; str = "hello"; p str.object_id; str; end
end
str2 = Java::Blah.blah(MyFoo.new)
p str2.object_id
```
output:
```txt
2000
2000
```

Are you sure you tried to use IRubyObject for the interface?

**Comment by headius** [comment]
Oh and since this was about proc/interface conversion:

```ruby
$ jruby -e 'str2 = Java::Blah.blah(->{str = "hello"; p str.object_id; str}); p str2.object_id'
2000
2000
```

**Comment by klobuczek** [comment]
@headius Thanks for the samples. However, the java interface I am using belongs to a library and looks like that https://neo4j.com/docs/api/java-driver/current/org/neo4j/driver/v1/TransactionWork.html
I would have to wrap all the relevant parts of the java library to return IRubyObject. I would prefer a solution where I don't have to write any java code. Is there any ruby wrapper which the auto-coercion would not touch?

**Comment by kares** [comment]
that is unfortunate, I am sure you could do some proxy style object wrapping to 'disable' the conversion.

would love to look into adding an option to disable - would be nice and maybe even become a default in a next major release, unfortunately its a lot of work. although that might be due my plans being too big :) as we would like to make a lot of Java types be duck-type-like Ruby ones when the conversion isn't there.
anyway, you should be able to come up with a fairly simple prototype of an opt disable the conversion ...

**Comment by headius** [comment]
This is still an open question: how to handle passing Ruby objects to Java without doing so much conversion? Currently we still depend on automatically converting "core" values like numbers and strings when calling into Java and when coming back. We might be able to avoid it when coming back, but if a Java API wants a String we need to pass it a String; when it wants an Object, but we pass a Ruby String, we probably should pass a Java String.

In any case there's no clear path forward on this so I am detargeting as a "post-9000" issue. If anyone has ideas about how to make our Java integration do less conversion while still working seamlessly, I would love to chat about it!
